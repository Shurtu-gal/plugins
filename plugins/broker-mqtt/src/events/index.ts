export * from "./createServerDotEnv";
export * from "./createPackageJson";
export * from "./createAppModule";
export * from "./createDockerCompose";
