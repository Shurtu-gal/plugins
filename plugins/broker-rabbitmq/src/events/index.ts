export * from "./createServer";
export * from "./createMessageBrokerService";
export * from "./createConnectMicroservices";
