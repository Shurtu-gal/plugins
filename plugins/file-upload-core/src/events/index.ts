export * from "./createPackageJson";
export * from "./createServerAppModule";
export * from "./createServer";
export * from "./createMainFile";
export * from "./createEntityControllerBase";
export * from "./createEntityModuleBase";
export * from "./createEnityResolverBase";
