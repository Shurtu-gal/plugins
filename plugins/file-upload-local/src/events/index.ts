export * from "./createServer";
export * from "./createServerGitIgnore";
export * from "./createEntityService";
