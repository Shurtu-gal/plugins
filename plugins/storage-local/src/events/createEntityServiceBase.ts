import {
  CreateEntityServiceBaseParams,
  DsgContext,
  EntityField,
  EnumDataType,
} from "@amplication/code-gen-types";
import {
  addImports,
  addInjectableDependency,
  getClassDeclarationById,
  importNames,
  interpolate,
} from "../util/ast";
import { STORAGE_SERVICE_ID, STORAGE_SERVICE_MEMBER_ID } from "../constants";
import { builders } from "ast-types";
import { camelCase } from "lodash";
import { pascalCase } from "pascal-case";

export const beforeCreateEntityServiceBase = async (
  context: DsgContext,
  eventParams: CreateEntityServiceBaseParams,
) => {
  const { entity, template, templateMapping, serviceBaseId } = eventParams;

  const fileFields = entity.fields.filter(
    (field) => field.dataType === EnumDataType.File,
  );

  if (fileFields.length === 0) {
    return eventParams;
  }

  interpolate(template, templateMapping);

  const classDeclaration = getClassDeclarationById(template, serviceBaseId);

  addInjectableDependency(
    classDeclaration,
    STORAGE_SERVICE_MEMBER_ID.name,
    STORAGE_SERVICE_ID,
    "protected",
  );

  const importArray = [
    importNames(
      [STORAGE_SERVICE_ID],
      "src/storage/providers/local/local.storage.service",
    ),
    importNames([builders.identifier("InputJsonValue")], "src/types"),
    importNames(
      [builders.identifier("FileDownload"), builders.identifier("FileUpload")],
      "src/storage/base/storage.types",
    ),
    importNames(
      [builders.identifier("LocalStorageFile")],
      "src/storage/providers/local/local.storage.types",
    ),
  ];

  addImports(template, importArray);

  fileFields.forEach((field) => {
    classDeclaration.body.body.push(
      getUploadFunction(entity.name, field),
      getDownloadFunction(entity.name, field.name),
      getDeleteFunction(entity.name, field.name),
    );
  });

  return eventParams;
};

const getUploadFunction = (entityName: string, field: EntityField) => {
  const fieldName = field.name;

  const { properties } = field;

  const pascalEntityName = builders.identifier(pascalCase(entityName));
  const pascalFieldName = builders.identifier(pascalCase(fieldName));
  const camelCaseEntityName = builders.identifier(camelCase(entityName));
  const camelCaseFieldName = builders.identifier(camelCase(fieldName));

  const args = builders.identifier("args");
  args.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(
      builders.identifier("Prisma.SelectSubset"),
      builders.tsTypeParameterInstantiation([
        builders.tsTypeReference(builders.identifier("T")),
        builders.tsTypeReference(
          builders.identifier(`Prisma.${pascalEntityName.name}FindUniqueArgs`),
        ),
      ]),
    ),
  );

  const file = builders.identifier("file");
  file.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(builders.identifier("FileUpload")),
  );

  const initialReassignmentStatement = [
    builders.expressionStatement(
      builders.assignmentExpression(
        "=",
        builders.memberExpression(
          builders.identifier("file"),
          builders.identifier("filename"),
        ),
        // `profilePicture-${args.where.id}.${file.filename.split(".").pop()}`;
        builders.templateLiteral(
          [
            builders.templateElement(
              {
                raw: "profilePicture-",
                cooked: "profilePicture-",
              },
              false,
            ),
            builders.templateElement(
              {
                raw: ".",
                cooked: ".",
              },
              false,
            ),
            builders.templateElement(
              {
                raw: "",
                cooked: "",
              },
              true,
            ),
          ],
          [
            builders.memberExpression(
              builders.memberExpression(
                builders.identifier("args"),
                builders.identifier("where"),
              ),
              builders.identifier("id"),
            ),
            builders.callExpression(
              builders.memberExpression(
                builders.memberExpression(
                  builders.memberExpression(
                    builders.identifier("file"),
                    builders.identifier("filename"),
                  ),
                  builders.callExpression(builders.identifier("split"), [
                    builders.stringLiteral("."),
                  ]),
                ),
                builders.identifier("pop"),
              ),
              [],
            ),
          ],
        ),
      ),
    ),
    builders.variableDeclaration("const", [
      builders.variableDeclarator(
        builders.identifier("containerPath"),
        builders.stringLiteral(properties?.containerPath || fieldName),
      ),
    ]),
  ];

  return builders.classMethod.from({
    async: true,
    kind: "method",
    key: builders.identifier(`upload${pascalFieldName.name}`),
    params: [args, file],
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(
        builders.identifier(`Promise<Prisma${pascalEntityName.name}>`),
      ),
    ),
    typeParameters: builders.tsTypeParameterDeclaration.from({
      params: [
        builders.tsTypeParameter.from({
          name: "T",
          constraint: builders.tsTypeReference(
            builders.identifier(
              `Prisma.${pascalEntityName.name}FindUniqueArgs`,
            ),
          ),
        }),
      ],
    }),
    body: builders.blockStatement([
      ...initialReassignmentStatement,
      builders.variableDeclaration("const", [
        builders.variableDeclarator(
          builders.identifier(camelCaseFieldName.name),
          builders.awaitExpression(
            builders.callExpression(
              builders.memberExpression(
                builders.memberExpression(
                  builders.thisExpression(),
                  builders.identifier("localStorageService"),
                ),
                builders.identifier("uploadFile"),
              ),
              [
                builders.identifier("file"),
                builders.arrayExpression(
                  properties?.allowedFileTypes?.map((mimeType: string) =>
                    builders.stringLiteral(mimeType),
                  ) || [],
                ),
                builders.numericLiteral(properties?.maxFileSize || 1000000),
                builders.identifier("containerPath"),
              ],
            ),
          ),
        ),
      ]),
      builders.returnStatement(
        builders.awaitExpression(
          builders.callExpression(
            builders.memberExpression(
              builders.memberExpression(
                builders.thisExpression(),
                builders.identifier("prisma"),
              ),
              builders.identifier(`${camelCaseEntityName.name}.update`),
            ),
            [
              builders.objectExpression([
                builders.property(
                  "init",
                  builders.identifier("where"),
                  builders.identifier("args.where"),
                ),
                builders.property(
                  "init",
                  builders.identifier("data"),
                  builders.objectExpression([
                    builders.property(
                      "init",
                      camelCaseFieldName,
                      builders.tsAsExpression(
                        camelCaseFieldName,
                        builders.tsTypeReference(
                          builders.identifier("InputJsonValue"),
                        ),
                      ),
                    ),
                  ]),
                ),
              ]),
            ],
          ),
        ),
      ),
    ]),
  });
};

const getDeleteFunction = (entityName: string, fieldName: string) => {
  const pascalEntityName = builders.identifier(pascalCase(entityName));
  const pascalFieldName = builders.identifier(pascalCase(fieldName));
  const camelCaseEntityName = builders.identifier(camelCase(entityName));
  const camelCaseFieldName = builders.identifier(camelCase(fieldName));

  const args = builders.identifier("args");
  args.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(
      builders.identifier("Prisma.SelectSubset"),
      builders.tsTypeParameterInstantiation([
        builders.tsTypeReference(builders.identifier("T")),
        builders.tsTypeReference(
          builders.identifier(`Prisma.${pascalEntityName.name}FindUniqueArgs`),
        ),
      ]),
    ),
  );

  return builders.classMethod.from({
    async: true,
    kind: "method",
    key: builders.identifier(`delete${pascalFieldName.name}`),
    params: [args],
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(
        builders.identifier(`Promise<Prisma${pascalEntityName.name}>`),
      ),
    ),
    typeParameters: builders.tsTypeParameterDeclaration.from({
      params: [
        builders.tsTypeParameter.from({
          name: "T",
          constraint: builders.tsTypeReference(
            builders.identifier(
              `Prisma.${pascalEntityName.name}FindUniqueArgs`,
            ),
          ),
        }),
      ],
    }),
    body: builders.blockStatement([
      builders.variableDeclaration("const", [
        builders.variableDeclarator(
          builders.objectPattern([
            builders.property.from({
              key: camelCaseFieldName,
              kind: "init",
              value: camelCaseFieldName,
              shorthand: true,
            }),
          ]),
          builders.awaitExpression(
            builders.callExpression(
              builders.memberExpression(
                builders.memberExpression(
                  builders.memberExpression(
                    builders.thisExpression(),
                    builders.identifier("prisma"),
                  ),
                  builders.identifier(camelCaseEntityName.name),
                ),
                builders.identifier("findUniqueOrThrow"),
              ),
              [
                builders.objectExpression([
                  builders.property(
                    "init",
                    builders.identifier("where"),
                    builders.identifier("args.where"),
                  ),
                ]),
              ],
            ),
          ),
        ),
      ]),
      builders.expressionStatement(
        builders.awaitExpression(
          builders.callExpression(
            builders.memberExpression(
              builders.memberExpression(
                builders.thisExpression(),
                builders.identifier("localStorageService"),
              ),
              builders.identifier("deleteFile"),
            ),
            [
              builders.tsAsExpression(
                builders.tsAsExpression(
                  camelCaseFieldName,
                  builders.tsUnknownKeyword(),
                ),
                builders.tsTypeReference(
                  builders.identifier("LocalStorageFile"),
                ),
              ),
            ],
          ),
        ),
      ),
      builders.returnStatement(
        builders.awaitExpression(
          builders.callExpression(
            builders.memberExpression(
              builders.memberExpression(
                builders.thisExpression(),
                builders.identifier("prisma"),
              ),
              builders.identifier(`${camelCaseEntityName.name}.update`),
            ),
            [
              builders.objectExpression([
                builders.property(
                  "init",
                  builders.identifier("where"),
                  builders.identifier("args.where"),
                ),
                builders.property(
                  "init",
                  builders.identifier("data"),
                  builders.objectExpression([
                    builders.property(
                      "init",
                      camelCaseFieldName,
                      builders.identifier("Prisma.DbNull"),
                    ),
                  ]),
                ),
              ]),
            ],
          ),
        ),
      ),
    ]),
  });
};

const getDownloadFunction = (entityName: string, fieldName: string) => {
  const pascalEntityName = builders.identifier(pascalCase(entityName));
  const pascalFieldName = builders.identifier(pascalCase(fieldName));
  const camelCaseEntityName = builders.identifier(camelCase(entityName));
  const camelCaseFieldName = builders.identifier(camelCase(fieldName));

  const args = builders.identifier("args");
  args.typeAnnotation = builders.tsTypeAnnotation(
    builders.tsTypeReference(
      builders.identifier("Prisma.SelectSubset"),
      builders.tsTypeParameterInstantiation([
        builders.tsTypeReference(builders.identifier("T")),
        builders.tsTypeReference(
          builders.identifier(`Prisma.${pascalEntityName.name}FindUniqueArgs`),
        ),
      ]),
    ),
  );

  return builders.classMethod.from({
    async: true,
    kind: "method",
    key: builders.identifier(`download${pascalFieldName.name}`),
    params: [args],
    returnType: builders.tsTypeAnnotation(
      builders.tsTypeReference(builders.identifier("Promise<FileDownload>")),
    ),
    typeParameters: builders.tsTypeParameterDeclaration.from({
      params: [
        builders.tsTypeParameter.from({
          name: "T",
          constraint: builders.tsTypeReference(
            builders.identifier(
              `Prisma.${pascalEntityName.name}FindUniqueArgs`,
            ),
          ),
        }),
      ],
    }),
    body: builders.blockStatement([
      builders.variableDeclaration("const", [
        builders.variableDeclarator(
          builders.objectPattern([
            builders.property.from({
              key: camelCaseFieldName,
              kind: "init",
              value: camelCaseFieldName,
              shorthand: true,
            }),
          ]),
          builders.awaitExpression(
            builders.callExpression(
              builders.memberExpression(
                builders.memberExpression(
                  builders.memberExpression(
                    builders.thisExpression(),
                    builders.identifier("prisma"),
                  ),
                  builders.identifier(camelCaseEntityName.name),
                ),
                builders.identifier("findUniqueOrThrow"),
              ),
              [
                builders.objectExpression([
                  builders.property(
                    "init",
                    builders.identifier("where"),
                    builders.identifier("args.where"),
                  ),
                ]),
              ],
            ),
          ),
        ),
      ]),
      builders.returnStatement(
        builders.awaitExpression(
          builders.callExpression(
            builders.memberExpression(
              builders.memberExpression(
                builders.thisExpression(),
                builders.identifier("localStorageService"),
              ),
              builders.identifier("downloadFile"),
            ),
            [
              builders.tsAsExpression(
                builders.tsAsExpression(
                  camelCaseFieldName,
                  builders.tsUnknownKeyword(),
                ),
                builders.tsTypeReference(
                  builders.identifier("LocalStorageFile"),
                ),
              ),
            ],
          ),
        ),
      ),
    ]),
  });
};
