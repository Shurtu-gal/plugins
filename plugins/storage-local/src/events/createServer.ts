import {
  CreateServerParams,
  DsgContext,
  ModuleMap,
} from "@amplication/code-gen-types";
import { join, resolve } from "path";
import { templatesPath } from "../constants";
import { print, readFile } from "@amplication/code-gen-utils";
import { getPluginSettings } from "../utils";
import { interpolate } from "../util/ast";
import { builders } from "ast-types";

export const beforeCreateServer = async (context: DsgContext) => {
  const { pluginInstallations } = context;
  if (
    !pluginInstallations.some(
      (plugin) => plugin.npm === "@amplication/plugin-storage-core",
    )
  ) {
    throw new Error(
      "The storage-core plugin must be installed for the storage-local plugin to function",
    );
  }
};

export const afterCreateServer = async (
  context: DsgContext,
  eventParams: CreateServerParams,
  modules: ModuleMap,
) => {
  const { fileBasePath } = getPluginSettings(context.pluginInstallations);

  const providerServicePath = resolve(
    templatesPath,
    "local.storage.service.template.ts",
  );
  const providerService = await readFile(providerServicePath);

  const templateMapping = {
    BASE_PATH: builders.stringLiteral(fileBasePath || "uploads"),
  };

  interpolate(providerService, templateMapping);

  const providerTypesPath = resolve(
    templatesPath,
    "local.storage.types.template.ts",
  );

  const providerTypes = await readFile(providerTypesPath);

  await modules.set({
    code: print(providerService).code,
    path: join(
      context.serverDirectories.srcDirectory,
      "storage",
      "providers",
      "local",
      "local.storage.service.ts",
    ),
  });

  await modules.set({
    code: print(providerTypes).code,
    path: join(
      context.serverDirectories.srcDirectory,
      "storage",
      "providers",
      "local",
      "local.storage.types.ts",
    ),
  });

  return modules;
};
