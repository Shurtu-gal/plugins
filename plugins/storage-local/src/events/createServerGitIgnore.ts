import {
  CreateServerGitIgnoreParams,
  DsgContext,
} from "@amplication/code-gen-types";
import { getPluginSettings } from "../utils";

export const beforeCreateServerGitIgnore = async (
  context: DsgContext,
  eventParams: CreateServerGitIgnoreParams,
) => {
  const { gitignorePaths } = eventParams;
  const { fileBasePath } = getPluginSettings(context.pluginInstallations);

  gitignorePaths.push(`/${fileBasePath || "uploads"}`);

  return eventParams;
};
