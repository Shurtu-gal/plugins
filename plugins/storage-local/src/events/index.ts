export * from "./createServer";
export * from "./createServerGitIgnore";
export * from "./createEntityService";
export * from "./createEntityServiceBase";
