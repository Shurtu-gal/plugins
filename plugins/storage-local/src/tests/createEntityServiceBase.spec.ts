import {
  CreateEntityServiceBaseParams,
  DsgContext,
  EnumDataType,
} from "@amplication/code-gen-types";
import { mock } from "jest-mock-extended";
import { builders } from "ast-types";
import { beforeCreateEntityServiceBase } from "../events";
import { print, readFile } from "@amplication/code-gen-utils";
import { resolve, join } from "path";

describe("Test createEntityServiceBase hook", () => {
  let context: DsgContext;
  let eventParams: CreateEntityServiceBaseParams;

  beforeEach(() => {
    context = mock<DsgContext>();
    eventParams = mock<CreateEntityServiceBaseParams>({
      entity: {
        id: "user",
        name: "User",
        displayName: "User",
        pluralDisplayName: "Users",
        fields: [
          {
            id: "profilePicture",
            name: "profilePicture",
            displayName: "Profile Picture",
            dataType: EnumDataType.File,
            properties: {
              allowedFileTypes: ["image/jpeg", "image/png"],
              containerPath: "profile-pictures",
              maxFileSize: 10485760,
            },
            required: false,
            searchable: false,
          },
        ],
      },
      serviceBaseId: builders.identifier("UserServiceBase"),
    });
  });

  it("should add all the necessary imports", async () => {
    eventParams.template = await readFile(
      resolve(join(__dirname, "./fixtures/service.base.template.ts")),
    );
    eventParams.templateMapping = {
      SERVICE_BASE: builders.identifier("UserServiceBase"),
    };

    const { template } = await beforeCreateEntityServiceBase(
      context,
      eventParams,
    );
    expect(print(template).code).toMatchSnapshot();
  });
});
