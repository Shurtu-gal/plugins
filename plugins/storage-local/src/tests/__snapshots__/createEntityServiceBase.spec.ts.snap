// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test createEntityServiceBase hook should add all the necessary imports 1`] = `
"import { PrismaService } from "../../prisma/prisma.service";
import { Prisma, ENTITY as PRISMA_ENTITY } from "@prisma/client";
import { LocalStorageService } from "src/storage/providers/local/local.storage.service";
import { InputJsonValue } from "src/types";
import { FileDownload, FileUpload } from "src/storage/base/storage.types";
import { LocalStorageFile } from "src/storage/providers/local/local.storage.types";

declare const CREATE_ARGS_MAPPING: Prisma.CREATE_ARGS;
declare const UPDATE_ARGS_MAPPING: Prisma.UPDATE_ARGS;

export class UserServiceBase {
  constructor(
    protected readonly prisma: PrismaService,
    protected readonly localStorageService: LocalStorageService
  ) {}

  async count(args: Omit<Prisma.COUNT_ARGS, "select">): Promise<number> {
    return this.prisma.DELEGATE.count(args);
  }

  async FIND_MANY_ENTITY_FUNCTION<T extends Prisma.FIND_MANY_ARGS>(
    args: Prisma.SelectSubset<T, Prisma.FIND_MANY_ARGS>
  ): Promise<PRISMA_ENTITY[]> {
    return this.prisma.DELEGATE.findMany<Prisma.FIND_MANY_ARGS>(args);
  }
  async FIND_ONE_ENTITY_FUNCTION<T extends Prisma.FIND_ONE_ARGS>(
    args: Prisma.SelectSubset<T, Prisma.FIND_ONE_ARGS>
  ): Promise<PRISMA_ENTITY | null> {
    return this.prisma.DELEGATE.findUnique(args);
  }
  async CREATE_ENTITY_FUNCTION<T extends Prisma.CREATE_ARGS>(
    args: Prisma.SelectSubset<T, Prisma.CREATE_ARGS>
  ): Promise<PRISMA_ENTITY> {
    return this.prisma.DELEGATE.create<T>(CREATE_ARGS_MAPPING);
  }
  async UPDATE_ENTITY_FUNCTION<T extends Prisma.UPDATE_ARGS>(
    args: Prisma.SelectSubset<T, Prisma.UPDATE_ARGS>
  ): Promise<PRISMA_ENTITY> {
    return this.prisma.DELEGATE.update<T>(UPDATE_ARGS_MAPPING);
  }
  async DELETE_ENTITY_FUNCTION<T extends Prisma.DELETE_ARGS>(
    args: Prisma.SelectSubset<T, Prisma.DELETE_ARGS>
  ): Promise<PRISMA_ENTITY> {
    return this.prisma.DELEGATE.delete(args);
  }

  async uploadProfilePicture<T extends Prisma.UserFindUniqueArgs>(args: Prisma.SelectSubset<T, Prisma.UserFindUniqueArgs>, file: FileUpload): Promise<PrismaUser> {
    file.filename = \`profilePicture-\${args.where.id}.\${file.filename.split(".").pop()}\`;
    const containerPath = "profile-pictures";
    const profilePicture = await this.localStorageService.uploadFile(file, ["image/jpeg", "image/png"], 10485760, containerPath);

    return await this.prisma.user.update({
      where: args.where,

      data: {
        profilePicture: profilePicture as InputJsonValue
      }
    });
  }

  async downloadProfilePicture<T extends Prisma.UserFindUniqueArgs>(args: Prisma.SelectSubset<T, Prisma.UserFindUniqueArgs>): Promise<FileDownload> {
    const {
      profilePicture
    } = await this.prisma.user.findUniqueOrThrow({
      where: args.where
    });

    return await this.localStorageService.downloadFile(profilePicture as unknown as LocalStorageFile);
  }

  async deleteProfilePicture<T extends Prisma.UserFindUniqueArgs>(args: Prisma.SelectSubset<T, Prisma.UserFindUniqueArgs>): Promise<PrismaUser> {
    const {
      profilePicture
    } = await this.prisma.user.findUniqueOrThrow({
      where: args.where
    });

    await this.localStorageService.deleteFile(profilePicture as unknown as LocalStorageFile);

    return await this.prisma.user.update({
      where: args.where,

      data: {
        profilePicture: Prisma.DbNull
      }
    });
  }
}
"
`;
